### ISSUE TITLE
Spark integration can cause YARN applications to crash if Spark context is explicitly stopped

### ISSUE THREAD

**Issue by mzheng-plaid** [OpenLineage/OpenLineage#2513]
Related to https://github.com/OpenLineage/OpenLineage/issues/2511

We have jobs that take a very long time to run `buildInputDatasets` (> 90s). This means the listener events are processed after `spark.stop()` (the Spark context is stopped in our application), which is problematic because it seems like the AM continues to request containers from YARN even after the Spark context is stopped. These new containers crash because they're unable to connect to the scheduler, which eventually cause YARN application to be marked as failed.

```
Caused by: org.apache.spark.rpc.RpcEndpointNotFoundException: Cannot find endpoint: spark://CoarseGrainedScheduler@{ip}:{port}
```

I believe this is because in [https://github.com/OpenLineage/OpenLineage/blob/987e5b806dc8bd6c5aab5f85c97af76a87[…]n/java/io/openlineage/spark/agent/OpenLineageSparkListener.java](https://github.com/OpenLineage/OpenLineage/blob/987e5b806dc8bd6c5aab5f85c97af76a87b1de0f/integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java#L83) a `SparkListenerSQLExecutionEnd` event is processed after the Spark context is stopped - I believe `createSparkSQLExecutionContext` is doing something undefined in (https://github.com/OpenLineage/OpenLineage/blob/987e5b806dc8bd6c5aab5f85c97af76a87[…]n/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java). I verified this was happening by adding log statements that printed whether the Spark context was stopped when receiving the event.

I'm not sure if this is defined behavior for the Spark context to be accessed after its stopped? After I skipped the event in `onOtherEvent` if the Spark context is stopped it no longer crashes trying to spin up new containers in some of our jobs.

```
+  private boolean isSparkContextStopped() {
+    return asJavaOptional(SparkSession.getDefaultSession()
+      .map(sparkContextFromSession)
+      .orElse(activeSparkContext))
+      .map(
+        ctx -> {
+          return ctx.isStopped();
+        })
+      .orElse(true); // If for some reason we can't get the Spark context, we assume it's stopped
+  }
+
   @Override
   public void onOtherEvent(SparkListenerEvent event) {
     if (isDisabled) {
       return;
     }
+    if (isSparkContextStopped()) {
+      log.warn("SparkContext is stopped, skipping event: {}", event.getClass());
+      return;
+    }
```

However I don't think this is sufficient to patch this, we still have jobs that take >200s to process events fail even with this patch. It seems like this is because the Spark context is stopped _during_ processing `onOtherEvent` so this check is insufficient.

Could https://openlineage.io/docs/integrations/spark/configuration/circuit_breaker/ be used to trigger a circuit breaker if the Spark context is stopped?

**Comment by d-m-h** [comment]
Yeah - can confirm, we're suffering from a similar issue. The listener is attempting to access information after the spark context has been stopped, which in our case is causing the job to fail and restart.

**Comment by mobuchowski** [comment]
@d-m-h are you also using Yarn?

**Comment by d-m-h** [comment]
Nope. Spark on K8s.

**Comment by mobuchowski** [comment]
Yeah, then I think any Spark context access needs to be guarded if we can't guarantee that any Spark context is available. The problem is that complete event might miss any necessary information, but this is preferable to failing or hanging.

**Comment by mobuchowski** [comment]
Just for additional information @d-m-h - this was happening before, right? It's not a new bug?

**Comment by pawel-big-lebowski** [comment]
I don't understand why this could be an outcome of some recent changes. If it was like this, we would be able to address it with an immediate fix. Based on the description provided, it seems like this is something we have forever. We're relying on a closed context and accessing it causes timeouts which cumulate for all the visitors and dataset builder. 

I think this could be addressed with:
 * implementing in java client `CompositeCircuitBreaker` which can contain many circuit breakers 
 * create `SparkContextStoppedCircuitBreaker` to be added by default. 

Secondly, `openlineage-spark` should always catch all the exceptions possible. We're doing this for dataset visitors and builders. We should do this as well in `ContextFactory` but I am not sure where exactly is the exception thrown and where to catch it.

**Comment by d-m-h** [comment]
> I think this could be addressed with:
>
> * implementing in java client CompositeCircuitBreaker which can contain many circuit breakers
> * create SparkContextStoppedCircuitBreaker to be added by default.

I don't understand how a circuit breaker would prevent the connector from triggering a request to YARN (or in my case, Kubernetes)

**Comment by pawel-big-lebowski** [comment]
@d-m-h circuit breaker would not prevent triggering yarn request. But it would prevent creating dozens of such request and it would stop execution for a hanging request once it detects a stopped context. 

It is still not clear to me how is this killing a job.

**Comment by pawel-big-lebowski** [comment]
@mzheng-plaid Are you able to provide an exception pointing to OpenLineage code resulting in `RpcEndpointNotFoundException`? Would love to see why isn't the exception caught?

**Comment by mzheng-plaid** [comment]
@pawel-big-lebowski the exception is _not_ in OpenLineage code - its an indirect effect of YARN thinking new containers need to be spun up after the Spark session is stopped because of OpenLineage, this exception is in the container startup code (it makes sense these can't start up because the scheduler is stopped). After enough containers crash on startup YARN thinks the application itself is failed.

The reason I think the circuit breaker will work is because it should always happen fast enough to prevent YARN from attempting to spin up new containers.

**Comment by mzheng-plaid** [comment]
Hi, any updates here?

**Comment by mobuchowski** [comment]
@mzheng-plaid https://github.com/OpenLineage/OpenLineage/pull/2535 was merged. However, while it might fix the immediate issue, I'm not sure it fixes the underlying problem - as you yourself correctly noticed:

>I feel like it should be more obvious that something is wrong (without actually failing the application), if in my setup something is taking 90s to process an event I'd rather OpenLineage do nothing until I can remediate. The partial data doesn't seem useful without these facets. Ideally I'd also prefer OpenLineage stop everything and terminate if the application itself were done - I definitely wouldn't want https://github.com/OpenLineage/OpenLineage/issues/2375 to happen and the infrastructure to remain up.

There are few tiers to answer that question. 
First, it's obvious that we shouldn't take that long to generate lineage. I think one of the possible causes is slow implementation of column level lineage or bugs when defining datasets like https://github.com/OpenLineage/OpenLineage/issues/2511
Second is whether we can do it: 
> Ideally I'd also prefer OpenLineage stop everything and terminate if the application itself were done
because, really, it can be miliseconds between finishing a job and calling `spark.stop()`.

I think one solution would be to refactor integration to 
1. generate _basic_ lineage information, that could be accessed just by getting LogicalPlan properties and not doing any advanced processing or network calls,
2. then do more potentially time consuming stuff: generate column-level lineage, ask external catalogs about table schemas and much more
3. if spark context stops somewhere after 1), we terminate 2) and send basic event generated from 1).

I don't think generally skipping emitting `COMPLETE` event at all, and misleading OL consumers that the application is still running is the desired solution - but limiting what we do then might be.

**Comment by mzheng-plaid** [comment]
Thanks for the detailed answer @mobuchowski , your suggestion on refactoring seems at a high level reasonable. Overall as a user, its hard for me to advocate for OpenLineage to be enabled on all our apps if there is even a 0.1% chance of it causing Spark job failure - its hard to justify that risk. For now, we're operating a fork with the Spark circuit breaker I pasted above to unblock usage.

re: refactoring integration timeline, this seems like it may take a while? What should other users do in the interim?

> I don't think generally skipping emitting COMPLETE event at all, and misleading OL consumers that the application is still running is the desired solution - but limiting what we do then might be.

This would also happen on any circuit breakers being triggered right?

**Comment by pawel-big-lebowski** [comment]
@mzheng-plaid Could you turn on `debugFacet` (see https://openlineage.io/docs/integrations/spark/configuration/spark_conf) for the job where building inputs take more than 90s? This should contain some short desc of logical plan and can be helpful to find visitor or dataset builder which has a bug. 

Would adding a timeout to an existing circuit breaker solve the issue of breaking others jobs? To me these two separate problems: (1) assure OL is not failing jobs (2) OL always emits termination events. I don't mind solving (1) while giving more time for (2).

**Comment by mzheng-plaid** [comment]
(@pawel-big-lebowski sorry missed email - I added a reproducible example in https://github.com/OpenLineage/OpenLineage/issues/2511#issuecomment-2031232924 for the slow code)

I liked your original suggestion of being able to compose circuit breakers together (seems more flexible), but I think adding a timeout that's configurable to existing circuit breakers would work too

**Comment by pawel-big-lebowski** [comment]
@mzheng-plaid thanks for comment (I realised the other issue later too). I like @mobuchowski approach.

For now I prepared timeout PR https://github.com/OpenLineage/OpenLineage/pull/2609 as a short term solution.
